consensus/p2p/proposer/proposer.go

```diff
@@ -0,0 +1,195 @@
+package proposal
+
+import (
+	"iter"
+
+	"github.com/NethermindEth/juno/consensus/types"
+	"github.com/starknet-io/starknet-p2pspecs/p2p/proto/common"
+	"github.com/starknet-io/starknet-p2pspecs/p2p/proto/consensus/consensus"
+	"google.golang.org/protobuf/proto"
+)
+
+const txBatchSize = 64 // TODO: make this configurable
+
+type seqNoCounter[V types.Hashable[H], H types.Hash, A types.Addr] struct {
+	streamID string
+	seqNo    uint64
+}
+
+// ToProposalStream converts a proposal to a stream of messages.
+func ToProposalStream[V types.Hashable[H], H types.Hash, A types.Addr](
+	adapter ProposerAdapter[V, H, A],
+	proposal types.Proposal[V, H, A],
+) (iter.Seq2[*consensus.StreamMessage, error], error) {
+	streamID, err := buildStreamID(proposal)
+	if err != nil {
+		return nil, err // TODO: log error
+	}
+
+	return func(yield func(*consensus.StreamMessage, error) bool) {
+		seqNoCounter := seqNoCounter[V, H, A]{
+			streamID: streamID,
+		}
+
+		if !yield(seqNoCounter.fromProposalInit(adapter, &proposal)) {
+			return
+		}
+
+		if !yield(seqNoCounter.fromBlockInfo(adapter, &proposal)) {
+			return
+		}
+
+		txStream, err := adapter.ProposalTransactions(proposal)
+		if err != nil {
+			yield(nil, err)
+			return
+		}
+
+		txBatch := make([]*consensus.ConsensusTransaction, 0, txBatchSize)
+		for tx := range txStream {
+			txBatch = append(txBatch, tx)
+			if len(txBatch) == txBatchSize {
+				if !yield(seqNoCounter.fromTransactions(txBatch)) {
+					return
+				}
+				txBatch = txBatch[:0]
+			}
+		}
+
+		if len(txBatch) > 0 {
+			if !yield(seqNoCounter.fromTransactions(txBatch)) {
+				return
+			}
+		}
```

Comment:
Just a suggestion to avoid the append overhead

```suggestion
		txBatch := make([]*consensus.ConsensusTransaction, txBatchSize)
		i := 0
		for tx := range txStream {
			txBatch[i] = tx
			i++

			if i == txBatchSize {
				if !yield(seqNoCounter.fromTransactions(txBatch[:i])) {
					return
				}
				i = 0
			}
		}

		if i > 0 {
			if !yield(seqNoCounter.fromTransactions(txBatch[:i])) {
				return
			}
		}
```

Reply:
I'm curious about this too, so I wrote a short benchmark to compare the performance between append with pre-allocated capacity and using an int as index:

```
package main

import (
	"iter"
	"math/rand"
	"testing"
)

const (
	maxSize    = 100000
	actualSize = maxSize / 2
)

type item struct {
	x, y, z int
}

func stream() iter.Seq[item] {
	return func(yield func(item) bool) {
		for range actualSize {
			if !yield(item{x: rand.Int(), y: rand.Int(), z: rand.Int()}) {
				return
			}
		}
	}
}

// Memory allocation benchmark
func BenchmarkDirectAssignmentMem(b *testing.B) {
	b.ReportAllocs()
	for range b.N {
		slice := make([]item, maxSize)
		stream := stream()
		j := 0
		for item := range stream {
			slice[j] = item
			j++
		}
	}
}

func BenchmarkAppendStrategyMem(b *testing.B) {
	b.ReportAllocs()
	for range b.N {
		slice := make([]item, 0, maxSize)
		stream := stream()
		for item := range stream {
			slice = append(slice, item)
		}
	}
}

func BenchmarkGrowingSliceMem(b *testing.B) {
	b.ReportAllocs()
	for range b.N {
		var slice []item
		stream := stream()
		for item := range stream {
			slice = append(slice, item)
		}
	}
}
```

![image](https://github.com/user-attachments/assets/da24c042-c840-447f-bf51-0b1c162bef25)

Reply:
That said, I think direct assignment can get slight performance gain because it doesn't have to check bound, while append has better clarity. I don't have strong opinion here, so let me know what you think.

Reply:
I think we've been trying to go for the direct assignment over the append approach recently, given it's more efficient